sys/linux/init_vusb.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package linux
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/google/syzkaller/prog"
+)
+
+const (
+	USB_DEVICE_ID_MATCH_VENDOR       = 0x0001
+	USB_DEVICE_ID_MATCH_PRODUCT      = 0x0002
+	USB_DEVICE_ID_MATCH_DEV_LO       = 0x0004
+	USB_DEVICE_ID_MATCH_DEV_HI       = 0x0008
+	USB_DEVICE_ID_MATCH_DEV_CLASS    = 0x0010
+	USB_DEVICE_ID_MATCH_DEV_SUBCLASS = 0x0020
+	USB_DEVICE_ID_MATCH_DEV_PROTOCOL = 0x0040
+	USB_DEVICE_ID_MATCH_INT_CLASS    = 0x0080
+	USB_DEVICE_ID_MATCH_INT_SUBCLASS = 0x0100
+	USB_DEVICE_ID_MATCH_INT_PROTOCOL = 0x0200
+	USB_DEVICE_ID_MATCH_INT_NUMBER   = 0x0400
+)
+
+func (arch *arch) generateUsbDeviceID(g *prog.Gen, typ0 prog.Type, old prog.Arg) (
+	arg prog.Arg, calls []*prog.Call) {
+	var matchFlags uint16
+	var idVendor uint16
+	var idProduct uint16
+	var bcdDeviceLo uint16
+	var bcdDeviceHi uint16
+	var bDeviceClass uint8
+	var bDeviceSubClass uint8
+	var bDeviceProtocol uint8
+	var bInterfaceClass uint8
+	var bInterfaceSubClass uint8
+	var bInterfaceProtocol uint8
+	var bInterfaceNumber uint8
+
+	base := usbIds[g.Rand().Intn(len(usbIds))]
+	p := bytes.NewBuffer([]byte(base))
+	binary.Read(p, binary.LittleEndian, &matchFlags)
+	binary.Read(p, binary.LittleEndian, &idVendor)
+	binary.Read(p, binary.LittleEndian, &idProduct)
+	binary.Read(p, binary.LittleEndian, &bcdDeviceLo)
+	binary.Read(p, binary.LittleEndian, &bcdDeviceHi)
+	binary.Read(p, binary.LittleEndian, &bDeviceClass)
+	binary.Read(p, binary.LittleEndian, &bDeviceSubClass)
+	binary.Read(p, binary.LittleEndian, &bDeviceProtocol)
+	binary.Read(p, binary.LittleEndian, &bInterfaceClass)
+	binary.Read(p, binary.LittleEndian, &bInterfaceSubClass)
+	binary.Read(p, binary.LittleEndian, &bInterfaceProtocol)
+	binary.Read(p, binary.LittleEndian, &bInterfaceNumber)
+
+	if (matchFlags & USB_DEVICE_ID_MATCH_VENDOR) == 0 {
+		idVendor = uint16(g.Rand().Intn(0xffff))
+	}
+	if (matchFlags & USB_DEVICE_ID_MATCH_PRODUCT) == 0 {
+		idProduct = uint16(g.Rand().Intn(0xffff))
+	}
+	if (matchFlags & USB_DEVICE_ID_MATCH_DEV_LO) == 0 {
+		bcdDeviceLo = 0x0
+	}
+	if (matchFlags & USB_DEVICE_ID_MATCH_DEV_HI) == 0 {
+		bcdDeviceHi = 0xffff
+	}
+	bcdDevice := bcdDeviceLo + uint16(g.Rand().Intn(int(bcdDeviceHi-bcdDeviceLo)))
```

Comment:
Shouldn't this be `bcdDeviceHi-bcdDeviceLo+1`?

Reply:
Fixed.